zephyr/wrapper.c

```diff
@@ -50,32 +50,85 @@ DECLARE_TR_CTX(zephyr_tr, SOF_UUID(zephyr_uuid), LOG_LEVEL_INFO);
 #define HEAP_SYSTEM_SIZE	0
 #endif
 
-/* The Zephyr heap - TODO: split heap */
-#define HEAP_SIZE	(HEAP_SYSTEM_SIZE + HEAP_RUNTIME_SIZE + HEAP_BUFFER_SIZE)
-static uint8_t __aligned(64) heapmem[HEAP_SIZE];
+/* The Zephyr heap */
+#define HEAPMEM_SIZE	(HEAP_SYSTEM_SIZE + HEAP_RUNTIME_SIZE + HEAP_BUFFER_SIZE)
+#define HEAPMEM_SHARED_SIZE	(HEAP_RUNTIME_SHARED_SIZE + HEAP_SYSTEM_SHARED_SIZE)
+
+static uint8_t heapmem[HEAPMEM_SIZE] __aligned(PLATFORM_DCACHE_ALIGN);
+static uint8_t heapmem_shared[HEAPMEM_SHARED_SIZE] __aligned(PLATFORM_DCACHE_ALIGN);
 
 /* Use k_heap structure */
 static struct k_heap sof_heap;
+static struct k_heap sof_heap_shared;
+
+static void *heapmem_addr(void)
+{
+	return uncache_to_cache(((void *)heapmem));
+}
 
 static int statics_init(const struct device *unused)
 {
 	ARG_UNUSED(unused);
 
-	sys_heap_init(&sof_heap.heap, heapmem, HEAP_SIZE);
+	sys_heap_init(&sof_heap.heap, heapmem_addr(), HEAPMEM_SIZE);
+	sys_heap_init(&sof_heap_shared.heap, heapmem_shared, HEAPMEM_SHARED_SIZE);
 
 	return 0;
 }
 
 SYS_INIT(statics_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_OBJECTS);
 
+static bool heap_is_cached(struct k_heap *h)
+{
+	return h == &sof_heap;
+}
+
+#define HEAP_CACHE_INVALIDATE true
+
+/*
+ * Flushes and optionally invalidates the whole heap area cache. Must
+ * be called with heap spinlock held.
+ */
+static void heap_cache_flush(struct k_heap *h, bool invalidate)
+{
+#if MP_NUM_CPUS > 1
+	if (heap_is_cached(h))
+		if (invalidate)
+			z_xtensa_cache_flush_inv(heapmem_addr(), HEAPMEM_SIZE);
+		else
+			z_xtensa_cache_flush(heapmem_addr(), HEAPMEM_SIZE);
+#endif
+}
+
 static void *heap_alloc_aligned(struct k_heap *h, size_t align, size_t bytes)
 {
 	void *ret = NULL;
 
 	k_spinlock_key_t key = k_spin_lock(&h->lock);
 
+	/*
+	 * Other cores may have accessed the heap, so invalidate the
+	 * cache.
+	 */
+	heap_cache_flush(h, HEAP_CACHE_INVALIDATE);
+
+	if (heap_is_cached(h)) {
+		/*
+		 * Ensure no heap metadata is mixed with allocated buffers by
+		 * aligning both start and size to dcache line size.
+		 */
+		__ASSERT(!(align & (PLATFORM_DCACHE_ALIGN - 1)), "alignment error");
+		bytes = ALIGN_UP(bytes, PLATFORM_DCACHE_ALIGN);
+	}
+
```

Comment:
@lyakh Ack, the previous version was not safe as caches of other cores may have held stale cache data of the heap still.

Here's an updated version that does the cache flushes in two phases, before and after all heap operations. Further, added a constraint to align-up the allocation sizes to ensure heap metadata is never put on the same cacheline as buffer payload data. This wouldn't be multicore safe.